### PR TITLE
Implement Protect integration tests

### PR DESCRIPTION
**Pull request by rukai**
closes https://github.com/shotover/shotover-proxy/issues/462

The motivation here is these integration tests will serve as replacements for the Protect unittests as those unittests are filled with implementation details of the old implementation.

I put the new tests in with the cassandra int tests as that is the only protocol that attempts to support protect at the moment.
When new protocols support it they should get their own int tests for it.

I added a new `protect` feature to allow us to write integration tests for protect while not including it in the public shotover API.
`#[cfg(test)]` does not allow this as it is only enabled during unit tests not integration tests.

I manually verified that the integration test tests the encryption code path as far as the current Protect implementation will go, but hopefully the message details rewrite will remove a lot of broken stuff and allow this test to actually test that encryption/decryption occurs.

### CODE REVIEW

**Comment by conorbros** [comment]
Why are we not including the Protect transform in the public API?

File: shotover-proxy/examples/cassandra-protect-local/docker-compose.yml
Comment (conorbros) [inline shotover-proxy/examples/cassandra-protect-local/docker-compose.yml:11]: Don't need this if you aren't using any UDFs

File: shotover-proxy/examples/cassandra-protect-aws/docker-entrypoint.sh
Comment (conorbros) [inline shotover-proxy/examples/cassandra-protect-aws/docker-entrypoint.sh:60]: Don't need this script if we aren't using any `CASSANDRA_*` env variables.

File: shotover-proxy/tests/cassandra_int_tests/basic_driver_tests.rs
Comment (conorbros) [inline shotover-proxy/tests/cassandra_int_tests/basic_driver_tests.rs:1144]: This could just be a function since it only has one test in it, unless there is more tests you are planning to add.

File: shotover-proxy/Cargo.toml
Comment (conorbros) [inline shotover-proxy/Cargo.toml:13]: I'm thinking this should be called `int-test` or similar so it's clear it has no purpose other than integration testing. Unless it's not just for integration testing?

Comment (rukai) [reply shotover-proxy/Cargo.toml:13]: In theory it would allow users who are willing to compile from source to try out new unreleased features.
Maybe we should call the feature `alpha-transforms` to allow for that case while also avoiding having lots of different features for each alpha transform floating around.

> Why are we not including the Protect transform in the public API?

New alpha transforms should not be public because:
* We dont want non-functional features to be available to users.
* We dont want to commit to a stable API yet.

File: shotover-proxy/examples/cassandra-protect-local/docker-compose.yml
Comment (rukai) [reply shotover-proxy/examples/cassandra-protect-local/docker-compose.yml:11]: I would like to run the full test suite in these tests eventually, but for now i've left the rest of the test suite out to make developing Protect easier.
Maybe I should just add the rest of the test suite now though?

File: shotover-proxy/tests/cassandra_int_tests/basic_driver_tests.rs
Comment (rukai) [reply shotover-proxy/tests/cassandra_int_tests/basic_driver_tests.rs:1144]: Having it as a module makes it easier to feature flag both the function and imports.

File: shotover-proxy/Cargo.toml
Comment (conorbros) [reply shotover-proxy/Cargo.toml:13]: That makes sense to me. I agree with the name suggestion `alpha-transforms`.

File: shotover-proxy/examples/cassandra-protect-local/docker-compose.yml
Comment (conorbros) [reply shotover-proxy/examples/cassandra-protect-local/docker-compose.yml:11]: Sure, might as well include in this PR.

Comment (rukai) [reply shotover-proxy/examples/cassandra-protect-local/docker-compose.yml:11]: alright, added

**Review by conorbros — APPROVED** [review]

File: shotover-proxy/examples/cassandra-protect-local/topology.yaml
Comment (benbromhead) [inline shotover-proxy/examples/cassandra-protect-local/topology.yaml:20]: I think yaml allows us to put the raw bytes in like a json array format which is a bit easier on the eyes `[ 30, 223, 12 ... ]`

**Review by benbromhead — APPROVED** [review]
One minor nit on the cassandra protect example yaml

File: shotover-proxy/examples/cassandra-protect-local/topology.yaml
Comment (rukai) [reply shotover-proxy/examples/cassandra-protect-local/topology.yaml:20]: Personally I would just leave it as is otherwise we have a 160 character wide line.
@conorbros which would you pick?

Comment (conorbros) [reply shotover-proxy/examples/cassandra-protect-local/topology.yaml:20]: I saw that and thought there would be a better way to do it (perhaps formatting it as hex?) but decided to leave it to not block this PR and get the tests in.

Comment (rukai) [reply shotover-proxy/examples/cassandra-protect-local/topology.yaml:20]: Alright, I kept it on one line and hexed the values.

Comment (conorbros) [reply shotover-proxy/examples/cassandra-protect-local/topology.yaml:20]: I meant as a long hex string.

Comment (conorbros) [reply shotover-proxy/examples/cassandra-protect-local/topology.yaml:20]: Not sure if it's possible to do that, but it seems strange to have to a supply a key byte by byte. I'd imagine you'd have it as a hex or base64 encoded string instead.

Comment (rukai) [reply shotover-proxy/examples/cassandra-protect-local/topology.yaml:20]: oh is it possible to write an array of integers as a hex string in yaml?

Comment (conorbros) [reply shotover-proxy/examples/cassandra-protect-local/topology.yaml:20]: I was thinking we read it as an encoded string and then convert into the bytes format that `sodiumoxide` expects.

Comment (rukai) [reply shotover-proxy/examples/cassandra-protect-local/topology.yaml:20]: ah right, I see what you mean.
That would certainly be nicer, but as you said earlier, lets address that later.

**Review by conorbros — APPROVED** [review]

File: shotover-proxy/examples/cassandra-protect-local/topology.yaml
Comment (conorbros) [reply shotover-proxy/examples/cassandra-protect-local/topology.yaml:20]: Opened an issue for it: #510

**Pull request merged**
